Profiler/Basics/BaseProfiler.cs

```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using NLog;
+using Profiler.Core;
+
+namespace Profiler.Basics
+{
+    /// <summary>
+    /// Implement basic features of a profiler that makes use of ProfilerEntry.
+    /// </summary>
+    /// <remarks>You can use ProfilerPatch without this class.</remarks>
+    public abstract class BaseProfiler<K> : IProfiler, IDisposable
+    {
+        // Holds onto ProfileResults until processed.
+        readonly ConcurrentQueue<ProfilerResult> _queuedProfilerResults;
+
+        /// <summary>
+        /// Cancels processing the ProfilerResult queue.
+        /// </summary>
+        readonly CancellationTokenSource _queueCanceller;
+
+        // Thread-safe dictionary of ProfilerEntry with an arbitrary type of keys.
+        readonly ConcurrentDictionary<K, ProfilerEntry> _profilerEntries;
+
+        // Cached function to unpool (or create) a new ProfilerEntity instance.
+        readonly Func<K, ProfilerEntry> _makeProfilerEntity;
+
+        bool _disposed;
+
+        protected BaseProfiler()
+        {
+            _queuedProfilerResults = new ConcurrentQueue<ProfilerResult>();
+            _queueCanceller = new CancellationTokenSource();
+            _profilerEntries = new ConcurrentDictionary<K, ProfilerEntry>();
+            _makeProfilerEntity = _ => ProfilerEntry.Pool.Instance.UnpoolOrCreate();
+        }
+
+        /// <inheritdoc/>
+        void IProfiler.OnProfileComplete(in ProfilerResult profilerResult)
+        {
+            _queuedProfilerResults.Enqueue(profilerResult);
```

Comment:
You should only have a single queue for all profilers.  Having each sample consumer in a separate thread, with a separate queue, adds a ton of complexity and overhead.

I'd suggest adding a single `ConcurrentQueue` to the `ProfilerPatch` that all the `StopToken` calls use, then use a single thread to consume those profiler results and dispatch them to all the registered observers.